### PR TITLE
refactor(streaming): use HashKey in join key type

### LINKED ISSUE #1772: bug(stream): incorrect `Row` equal comparison in case null value exist.
Currently, the equal comparison for `Row` is auto derived. However, since we are using `None` to represent the null value, an equal comparison between two identical rows with null values will produce `True`. Here null value (_unknow_) should be treated as false.
```rust
#[derive(Clone, Debug, Default, PartialEq, Eq, Hash)]
pub struct Row(pub Vec<Datum>);

pub type Datum = Option<ScalarImpl>;
```

cc. @TennyZhuang

### PR DESCRIPTION
**Pull request by yuhao-su**
## What's changed and what's your intention?

- use `HashKey` in join key type
- fix a bug when row contains a null datum.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
close #1772

### CODE REVIEW

**Review by liurenjie1024 — COMMENTED** [review]
LGTM

File: e2e_test/streaming/join.slt [hunk 1]
```diff
@@ -167,3 +193,9 @@ drop table t9
 
 statement ok
 drop table t10
+
+statement ok
+drop table t11
+
+statement ok
+drop table t12
```
Comment (TennyZhuang) [inline e2e_test/streaming/join.slt:201]:
```suggestion
drop table t12

```

File: src/common/src/collection/evictable.rs [hunk 2]
```diff
@@ -13,24 +13,31 @@
...
     /// Create a [`EvictableHashMap`] with the given target capacity.
-    pub fn new(target_cap: usize) -> Self {
+    pub fn new(target_cap: usize) -> EvictableHashMap<K, V> {
+        EvictableHashMap::new_with_hasher(target_cap, DefaultHasher::new())
+    }
+}
+
+impl<K: Hash + Eq, V, S: BuildHasher> EvictableHashMap<K, V, S> {
+    /// Create a [`EvictableHashMap`] with the given target capacity and haser.
+    pub fn new_with_hasher(target_cap: usize, hasher: S) -> Self {
```
Comment (TennyZhuang) [inline src/common/src/collection/evictable.rs:38]: `with_hasher` is recommended, see https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#general-conventions-[rfc-#430]

**Review by TennyZhuang — APPROVED** [review]
LGTM

File: src/common/src/collection/evictable.rs (hunk 2 above)
Comment (yuhao-su) [reply src/common/src/collection/evictable.rs]: Also we have other new_with_xxx in our repo. May fix those later.

**Pull request merged**